File: silverlocklib/group.cpp
#include "group.h"
#include "entry.h"

/*!
    \class Group
    Represents a group of password entries in a database.

    The parent of a group cannot be \c NULL and an assertion failure will
    result if a \c NULL pointer is passed to any of the constructors.
 */

Group::Group(const QString &title, Group *parent) :
    DatabaseNode(title)
{
    this->setParentNode(parent);
}

/*!
    Destroys the group, deleting all its child objects.

    The group's entries and subgroups are deleted upon deconstruction, which will recursively
    delete all subgroups and all entries contained within them. The group also removes itself
    from its parent group, if any.
 */
Group::~Group()
{
    this->detach();

    foreach (Entry *entry, this->entries())
    {
        delete entry;
    }

    foreach (Group *subgroup, this->groups())
    {
        delete subgroup;
    }
}

const QList<Group*>& Group::groups() const
{
    return this->m_groups;
}

const QList<Entry*>& Group::entries() const
{
    return this->m_entries;
}

/*!
    Gets the total number of groups in the group node, including subgroups.

    This method recursively queries each group for its number of subgroups, starting with but not
    counting the current group. If you want to get the number of groups in the current group only
    (non-recursive), call \link groups().count() \endlink.

    \sa countEntries()
 */
int Group::countGroups() const
{
    int total = 0;

    foreach (Group *group, this->groups())
    {
        total += group->countGroups();
    }

    return total + this->groups().count();
}

/*!
    Gets the total number of entries in the group node, including those in subgroups.

    This method recursively queries each group for its number of entries, starting with but not
    counting the current group. If you want to get the number of entries in the current group only
    (non-recursive), call \link entries().count() \endlink.

    \sa countGroups()
 */
int Group::countEntries() const
{
    int total = 0;

    foreach (Group *group, this->groups())
    {
        total += group->countEntries();
    }

    return total + this->entries().count();
}

/*!
    Determines whether the database contains a group or entry identified by the specified UUID.

    This method returns \c true if a group or entry exists in this database that is identified by
    \a uuid. If no such group or entry exists, this method returns \c false.

    \param uuid The UUID to search for.
 */
bool Group::containsUuid(const QUuid &uuid) const
{
    return this->isGroup(uuid) || this->isEntry(uuid);
}

/*!
    Determines whether the database contains a group identified by the specified UUID.

    This method returns \c true if a group exists in this database that is identified by \a uuid.
    If no such group exists, or an object exists but is not a child or descendant of the root group
    of this database (or is an entry), this method returns \c false.

    \param uuid The UUID to search for.
 */
bool Group::isGroup(const QUuid &uuid) const
{
    return this->findGroup(uuid) != NULL;
}

/*!
    Determines whether the database contains an entry identified by the specified UUID.

    This method returns \c true if an entry exists in this database that is identified by \a uuid.
    If no such entry exists, or an object exists but is not a child or descendant of the root group
    of this database (or is a group), this method returns \c false.

    \param uuid The UUID to search for.
 */
bool Group::isEntry(const QUuid &uuid) const
{
    return this->findEntry(uuid) != NULL;
}

/*!
    Retrieves a pointer to the group contained in this group node with the specified UUID.

    The recursive search starts from the group specified by \a startingGroup. If \a startingGroup
    is not contained within this database, a \c NULL pointer is returned.

    If no group identified by \a uuid exists, a \c NULL pointer is returned.

    \param uuid The UUID to search for.
    \param includeThis Whether to include this group node in the search. If \a uuid is equal to
    the UUID of this group node, the group node itself will be returned. Otherwise TODO TODO
 */
Group* Group::findGroup(const QUuid &uuid, bool includeThis)
{
    if (includeThis && this->uuid() == uuid)
    {
        return this;
    }
    else
    {
        return this->findGroupInternal(uuid);
    }
}

const Group* Group::findGroup(const QUuid &uuid, bool includeThis) const
{
    if (includeThis && this->uuid() == uuid)
    {
        return this;
    }
    else
    {
        return this->findGroupInternal(uuid);
    }
}

Group* Group::findGroupInternal(const QUuid &uuid) const
{
    foreach (Group *group, this->groups())
    {
        // If this group is a match, return it
        if (group->uuid() == uuid)
        {
            return group;
        }
        else
        {
            // Otherwise search its children
            Group *candidate = group->findGroup(uuid, true);
            if (candidate && candidate->uuid() == uuid)
            {
                return candidate;
            }
        }
    }

    return NULL;
}

/*!
    Retrieves a pointer to the entry contained in this group node with the specified UUID.

    The recursive search starts from the group specified by \a startingGroup. If \a startingGroup
    is not contained within this database, a \c NULL pointer is returned.

    If no entry identified by \a uuid exists, a \c NULL pointer is returned.

    \param uuid The UUID to search for.
 */
Entry* Group::findEntry(const QUuid &uuid) const
{
    foreach (Entry *entry, this->entries())
    {
        if (entry->uuid() == uuid)
        {
            return entry;
        }
    }

    foreach (Group *group, this->groups())
    {
        Entry *candidate = group->findEntry(uuid);
        if (candidate && candidate->uuid() == uuid)
        {
            return candidate;
        }
    }

    return NULL;
}

QList<Entry*> Group::findEntries(const SearchParameters &params) const
{
    if (!params.fieldsSelected())
    {
        return QList<Entry*>();
    }

    QRegExp regex = QRegExp(params.searchPattern, params.caseSensitiveEnum());
    QList<Entry*> entries;

    foreach (Entry *entry, this->entries())
    {
        QList<QString> fields = params.getDataList(*entry);
        foreach (QString field, fields)
        {
            if (params.useRegex)
            {
                if (!entries.contains(entry) && field.contains(regex))
                {
                    entries.append(entry);

                    // Break out of fields loop so we go to the next entry
                    break;
                }
            }
            else
            {
                if (!entries.contains(entry) && field.contains(params.searchPattern, params.caseSensitiveEnum()))
                {
                    entries.append(entry);

                    // Break out of fields loop so we go to the next entry
                    break;
                }
            }
        }
    }

    foreach (Group *group, this->groups())
    {
        entries.append(group->findEntries(params));
    }

    return entries;
}

void Group::attachToList()
{
    this->parentNode()->m_groups.append(this);
}

void Group::detachFromList()
{
    this->parentNode()->m_groups.removeAll(this);
}

QDomElement Group::toXml(QDomDocument &document) const
{
    QDomElement element = DatabaseNode::toXml(document);
    element.setTagName(XML_GROUP);
    return element;
}

File: src/lib/objc/Group.m
//
//  Group.m
//  Silverlock
//
//  Created by Jake Petroules on 1/27/11.
//  Copyright 2011 Petroules Corporation. All rights reserved.
//

#import "Group.i.h"
#import "DatabaseNode.i.h"

@interface Group (Private)

-(Group*) findGroupInternal: (CFUUIDRef) uuid;

@end

@implementation Group

-(id) init
{
	return [self initWithTitle: [NSString string]];
}

-(id) initWithTitle: (NSString*) title
{
	return [self initWithTitle: title andParent: nil];
}

-(id) initWithTitle: (NSString*) title andParent: (Group*) parent
{
	if (self = [super initWithTitle: title])
	{
		[self setParentNode: parent];
	}
	
	return self;
}

-(void) dealloc
{
	[self detach];
	
	NSArray *entries = [self entries];
	for (Entry *entry in entries)
	{
		[entry release];
	}
	
	NSArray *groups = [self groups];
	for (Group *subgroup in groups)
	{
		[subgroup release];
	}
	
	[super dealloc];
}

-(NSArray*) groups
{
	return [NSArray arrayWithArray: self->m_groups];
}

-(NSArray*) entries
{
	return [NSArray arrayWithArray: self->m_entries];
}

-(int) countGroups
{
	int total = 0;
	
	for (Group *group in self->m_groups)
	{
		total += [group countGroups];
	}
	
	return total + [self->m_groups count];
}

-(int) countEntries
{
	int total = 0;
	
	for (Group *group in self->m_groups)
	{
		total += [group countEntries];
	}
	
	return total + [self->m_entries count];
}

-(BOOL) containsUuid: (CFUUIDRef) uuid
{
	return [self isGroup: uuid] || [self isEntry: uuid];
}

-(BOOL) isGroup: (CFUUIDRef) uuid
{
	return [self findGroup: uuid] != nil;
}


-(BOOL) isEntry: (CFUUIDRef) uuid
{
	return [self findEntry: uuid] != nil;
}

-(Group*) findGroup: (CFUUIDRef) uuid
{
	return [self findGroup: uuid includeThis: NO];
}

-(Group*) findGroup: (CFUUIDRef) uuid includeThis: (BOOL) includeThis
{
	if (includeThis && CFEqual([self uuid], uuid))
	{
		return self;
	}
	else
	{
		return [self findGroupInternal: uuid];
	}
}

-(Entry*) findEntry: (CFUUIDRef) uuid
{
	for (Entry *entry in self->m_entries)
	{
		if (CFEqual([entry uuid], uuid))
		{
			return entry;
		}
	}
	
	for (Group *group in self->m_groups)
	{
		Entry *candidate = [group findEntry: uuid];
		if (candidate && CFEqual([candidate uuid], uuid))
		{
			return candidate;
		}
	}
	
	return nil;
}

@end

@implementation Group (Protected)

-(void) attachToList
{
	[[self parentNode]->m_groups addObject: self];
}

-(void) detachFromList
{
	[[self parentNode]->m_groups removeObjectIdenticalTo: self];
}

-(NSMutableArray*) internalGroupsList
{
	return self->m_groups;
}

-(NSMutableArray*) internalEntriesList
{
	return self->m_entries;
}

@end

@implementation Group (Private)

-(Group*) findGroupInternal: (CFUUIDRef) uuid
{
	for (Group *group in self->m_groups)
	{
		// If this group is a match, return it
		if (CFEqual([group uuid], uuid))
		{
			return group;
		}
		else
		{
			// Otherwise search its children
			Group *candidate = [group findGroup: uuid includeThis: YES];
			if (candidate && CFEqual([candidate uuid], uuid))
			{
				return candidate;
			}
		}
	}
	
	return nil;
}

@end